### DISCUSSION TITLE
I2C SSD1306 OLED with reset line: Wemos S2 Pico weirdness

### DISCUSSION THREAD

**Discussion by scruss** [micropython/micropython#10820] [Hardware & Peripherals] [answered]
SPI OLED screens with a reset line are pretty common, but the MicroPython [ssd1306](https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/display/ssd1306) library doesn't have any way to use them with I2C screens.

The Wemos [S2 Pico](https://www.wemos.cc/en/latest/s2/s2_pico.html) is an ESP32-S2 board with a 128x32 SSD1306 OLED built in. For the longest time, I couldn't get the display to work. The i2c bus scan would show the right device number, but attempting to display anything created noise:

![pico_garble](https://user-images.githubusercontent.com/425706/220787182-d758f171-37aa-4347-8c03-b0b34dcaf36a.jpg)

Looking at the [schematic](https://www.wemos.cc/en/latest/_static/files/sch_s2_pico_v1.0.0.pdf), apart from the nicely-documented pull-up resistors, there's an unexpected RES# line:

![Screenshot 2023-02-22 at 18-17-26 Printing —È - sch_s2_pico_v1 0 0 pdf](https://user-images.githubusercontent.com/425706/220787553-acca8624-6fe0-4ca1-8a4f-8ff7f3c4c28b.png)

It seems you have to set this line high for the display to work at all:

    oled_reset = Pin(18, Pin.OUT, value=1)

Seeing an I2C display with a reset line is a first for me. I wouldn't suggest adding support to the library as it can be handled well without changes.

**Comment by scruss** [comment, marked as answer]
Here's the complete code, including the required reset line activation, to make the OLED on the S2 Pico work:

``` python3
# for Wemos S2 Pico SSD1306 128x32 display
# https://www.wemos.cc/en/latest/s2/s2_pico.html
# scruss, 2023-02

from machine import Pin, I2C
import ssd1306

# ensure display reset is held high, otherwise nothing will work
oled_reset = Pin(18, Pin.OUT, value=1)
# ESP32 S2 only has one hardware I2C channel, #0 on pins 8 & 9
i2c = I2C(0, sda=Pin(8), scl=Pin(9))
# use default address 0x3C (60) for SSD1306
display = ssd1306.SSD1306_I2C(128, 32, i2c)
display.fill(0)
display.text("* Hello from *", 0, 0, 1)
display.text("*  S2 Pico   *", 0, 8, 1)
display.show()
```

**Reply by dnguyen76** [reply]
Do you have a solution for O.66" OLED 64 x 48, it seems don't have reset pin but it doesn't work

**Reply by scruss** [reply]
It seems to work for me. This for a Lolin [S2 mini](https://www.wemos.cc/en/latest/s2/s2_mini.html):

```python3
from machine import Pin, I2C, RTC
import s2mini
import ssd1306
import framebuf
import time
WIDTH = 64
HEIGHT = 48
rtc=RTC()

i2c=I2C(0, scl=Pin(s2mini.I2C_SCL), sda=Pin(s2mini.I2C_SDA))
print(i2c.scan())
# should return 49 (0x31 - I2C buttons) and 60 (0x3c - display)
oled = ssd1306.SSD1306_I2C(WIDTH, HEIGHT, i2c)

while True:
    oled.fill(0)
    oled.text("%02d:%02d:%02d" % rtc.datetime()[4:7], 0, 0)
    oled.show()
    time.sleep_ms(200)
```

**Comment by dnguyen76** [comment]
My configuration is Wemos D1 mini clone and OLED 64x48 0.66" shield

It seems i get a I2C problem as i2c.scan() function return 112 i2c peripherals from 0x08 to 0x77 ???
(if the oled shield is not connected  i2c.scan() return a timeout error)

i2c.readfrom(0x3C, 16)   return   0x3C :  b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  (same for 0x3D)
I had found a schematic from LOLIN but my OLED 64x48 shield is a clone
But i don't seen in schematic any information to solve this problem

any idea ?

[sch_oled_v2.1.0.pdf](https://github.com/micropython/micropython/files/11059675/sch_oled_v2.1.0.pdf)

**Comment by robert-hh** [comment]
It seems that you do not have pull-up resistors on sda and scl

**Comment by PBrunot** [comment]
It might be necessary to toggle the reset line. On my Lolin S2 Pico the display was still showing noise until I changed your code as follow:

```python
from machine import Pin, I2C
import ssd1306

oled_reset = Pin(18, Pin.OUT)
oled_reset.off()
oled_reset.on()

# ESP32 S2 only has one hardware I2C channel, #0 on pins 8 & 9
i2c = I2C(0, sda=Pin(8), scl=Pin(9))
# use default address 0x3C (60) for SSD1306
display = ssd1306.SSD1306_I2C(128, 32, i2c)
display.fill(0)
display.text("* Hello from *", 0, 0, 1)
display.text("*  S2 Pico   *", 0, 8, 1)
display.show()
```

**Reply by scruss** [reply]
The `oled_reset = Pin(18, Pin.OUT, value=1)` didn't work?

Note: I wasn't really looking for further discussion on this point. This platform doesn't work well for noting observations that don't require further input.

**Reply by PBrunot** [reply]
I confirm it did not work as it is, toggling the bit was necessary. 
I had just flashed the S2 Pico to Micropython 1.20, not sure if it matters.
I replied here because I reached this discussion through Google Search. It came pretty high on "S2 Pico SSD noise issue" (or something like that), so I thought it might be helpful for more people.

**Reply by frazzmatazz** [reply]
I also confirm it didn't work (just snow/noise) until I used PBrunot's code.

**Comment by scruss** [comment]
Note: I wasn't really looking for further discussion on this point. This platform doesn't work well for noting observations that don't require further input.

**Comment by D0NM** [comment]
thank you all, 
i revived OLED on this dusty LOLIN S2 PICO

#define OLED_RESET_PIN 18

void setup() {
  pinMode(OLED_RESET_PIN, OUTPUT);
  digitalWrite(OLED_RESET_PIN, LOW);
  digitalWrite(OLED_RESET_PIN, HIGH);
  ...
